### PR TITLE
Investment Transfers, Operations, and Currency Guardrails

### LINKED ISSUE #18: Create Investment History & Operations Tracking
**Description:**  
Create comprehensive investment history tracking and operations logging to provide audit trails and enable accurate profit/loss calculations. Integrate with transaction system so that investment transactions automatically create operation records. Based on previous investment tracker design, adapted for Ash framework and ash_money integration.

**Acceptance Criteria:**
- [ ] Create `Moolah.Finance.InvestmentHistory` resource
- [ ] Create `Moolah.Finance.Operation` resource  
- [ ] Add automatic history creation when investments are created/updated
- [ ] Track three operation types: deposit, withdraw, update
- [ ] Add fields with proper money handling:
  - InvestmentHistory: `value` (money), `investment_id`, timestamps
  - Operation: `type` (enum), `value` (money), `investment_id`, timestamps
- [ ] Create relationships between Investment, InvestmentHistory, and Operation
- [ ] Add Ash actions and changes to automatically create history records
- [ ] Ensure currency consistency across all history records
- [ ] Add integration with Transaction system (Issue #5)
- [ ] Create operations when investment transactions are created
- [ ] Add logic to determine which investment gets the operation
- [ ] Write comprehensive tests for history tracking and operations
- [ ] Add database migrations with proper indexes
- [ ] Update Investment resource to integrate with history tracking

**Example Database Schema:**

Note: Feel free to use ecto or other types of migrations, rather than just writing pure sql
```sql
-- Investment History - snapshots of investment values over time
CREATE TABLE investment_histories (
  id UUID PRIMARY KEY DEFAULT uuid_generate_v7(),
  value MONEY_WITH_CURRENCY NOT NULL,
  investment_id UUID NOT NULL REFERENCES investments(id) ON DELETE CASCADE,
  inserted_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
  updated_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW()
);

-- Operations - specific actions performed on investments  
CREATE TABLE operations (
  id UUID PRIMARY KEY DEFAULT uuid_generate_v7(),
  type VARCHAR(10) NOT NULL CHECK (type IN ('deposit', 'withdraw', 'update')),
  value MONEY_WITH_CURRENCY NOT NULL,
  investment_id UUID NOT NULL REFERENCES investments(id) ON DELETE CASCADE,
  inserted_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
  updated_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW()
);

-- Indexes for efficient querying
CREATE INDEX idx_investment_histories_investment_id ON investment_histories(investment_id);
CREATE INDEX idx_investment_histories_inserted_at ON investment_histories(inserted_at);
CREATE INDEX idx_operations_investment_id ON operations(investment_id);
CREATE INDEX idx_operations_type ON operations(type);
CREATE INDEX idx_operations_inserted_at ON operations(inserted_at);
```

** Example InvestmentHistory Resource:**
```elixir
defmodule Moolah.Finance.InvestmentHistory do
  use Ash.Resource,
    domain: Moolah.Finance,
    data_layer: AshPostgres.DataLayer

  postgres do
    table "investment_histories"
    repo Moolah.Repo
  end

  actions do
    defaults [:read]
    
    create :create do
      accept [:value, :investment_id]
      
      validate {Moolah.Finance.Validations.ValidateHistoryCurrency, []}
    end
  end

  attributes do
    uuid_v7_primary_key :id
    
    attribute :value, :money do
      allow_nil? false
      constraints storage_type: :money_with_currency
    end
    
    timestamps()
  end

  relationships do
    belongs_to :investment, Moolah.Finance.Investment do
      attribute_writable? true
      allow_nil? false
    end
  end
end
```

**Example Operation Resource:**
```elixir  
defmodule Moolah.Finance.Operation do
  use Ash.Resource,
    domain: Moolah.Finance,
    data_layer: AshPostgres.DataLayer

  @types [:deposit, :withdraw, :update]

  postgres do
    table "operations"
    repo Moolah.Repo
  end

  actions do
    defaults [:read]
    
    create :create do
      accept [:type, :value, :investment_id]
      
      validate {Moolah.Finance.Validations.ValidateOperationCurrency, []}
    end
  end

  attributes do
    uuid_v7_primary_key :id
    
    attribute :type, :atom do
      constraints one_of: @types
      allow_nil? false
    end
    
    attribute :value, :money do
      allow_nil? false
      constraints storage_type: :money_with_currency
    end
    
    timestamps()
  end

  relationships do
    belongs_to :investment, Moolah.Finance.Investment do
      attribute_writable? true
      allow_nil? false
    end
  end
end
```

**Example Updated Investment Resource (add relationships):**
```elixir
# Add to Issue #6a Investment resource:

relationships do
  belongs_to :account, Moolah.Ledger.Account do
    attribute_writable? true
    allow_nil? false
  end
  
  # Add these new relationships:
  has_many :investment_histories, Moolah.Finance.InvestmentHistory do
    destination_attribute :investment_id
  end
  
  has_many :operations, Moolah.Finance.Operation do
    destination_attribute :investment_id
  end
end
```

**Example Automatic History Creation:**
```elixir
# Add to Investment resource actions in Issue #6a:

create :create do
  accept [:name, :type, :subtype, :initial_value, :current_value, 
          :expiration_date, :account_id]
  
  validate {Moolah.Finance.Validations.ValidateSubtype, []}
  validate {Moolah.Finance.Validations.ValidateInvestmentCurrency, []}
  
  # Add automatic history creation
  change after_action(fn changeset, investment ->
    # Create initial history record
    Moolah.Finance.InvestmentHistory
    |> Ash.Changeset.for_create(:create, %{
      value: investment.current_value,
      investment_id: investment.id
    })
    |> Ash.create!()
    
    {:ok, investment}
  end)
end

update :update do
  accept [:name, :current_value, :expiration_date]
  
  validate {Moolah.Finance.Validations.ValidateInvestmentCurrency, []}
  
  # Add automatic history creation on updates
  change before_action(fn changeset ->
    old_value = changeset.data.current_value
    new_value = Ash.Changeset.get_attribute(changeset, :current_value)
    
    # Store old value for comparison
    Ash.Changeset.put_context(changeset, :old_value, old_value)
  end)
  
  change after_action(fn changeset, investment ->
    old_value = Ash.Changeset.get_context(changeset, :old_value)
    new_value = investment.current_value
    
    # Only create history if value actually changed
    if old_value != new_value do
      # Create history record
      Moolah.Finance.InvestmentHistory
      |> Ash.Changeset.for_create(:create, %{
        value: new_value,
        investment_id: investment.id
      })
      |> Ash.create!()
      
      # Create operation record
      value_change = Money.subtract(new_value, old_value)
      operation_type = if Money.positive?(value_change), do: :deposit, else: :withdraw
      
      Moolah.Finance.Operation
      |> Ash.Changeset.for_create(:create, %{
        type: operation_type,
        value: Money.abs(value_change),
        investment_id: investment.id
      })
      |> Ash.create!()
    end
    
    {:ok, investment}
  end)
end
```

**Example Currency Validation Modules:**
```elixir
defmodule Moolah.Finance.Validations.ValidateHistoryCurrency do
  use Ash.Resource.Validation
  
  @impl true
  def validate(changeset, _opts) do
    investment_id = Ash.Changeset.get_attribute(changeset, :investment_id)
    value = Ash.Changeset.get_attribute(changeset, :value)
    
    with {:ok, investment} <- Moolah.Finance.Investment |> Ash.get(investment_id),
         {:ok, value_currency} <- get_currency(value),
         {:ok, investment_currency} <- get_currency(investment.current_value) do
      
      if value_currency == investment_currency do
        :ok
      else
        {:error, field: :value, message: "currency must match investment currency (#{investment_currency})"}
      end
    else
      _ -> {:error, field: :investment_id, message: "invalid investment"}
    end
  end
  
  defp get_currency(%Money{currency: currency}), do: {:ok, to_string(currency)}
  defp get_currency(_), do: {:error, :invalid_money}
end

defmodule Moolah.Finance.Validations.ValidateOperationCurrency do
  use Ash.Resource.Validation
  
  @impl true  
  def validate(changeset, _opts) do
    investment_id = Ash.Changeset.get_attribute(changeset, :investment_id)
    value = Ash.Changeset.get_attribute(changeset, :value)
    
    with {:ok, investment} <- Moolah.Finance.Investment |> Ash.get(investment_id),
         {:ok, value_currency} <- get_currency(value),
         {:ok, investment_currency} <- get_currency(investment.current_value) do
      
      if value_currency == investment_currency do
        :ok
      else
        {:error, field: :value, message: "currency must match investment currency (#{investment_currency})"}
      end
    else
      _ -> {:error, field: :investment_id, message: "invalid investment"}
    end
  end
  
  defp get_currency(%Money{currency: currency}), do: {:ok, to_string(currency)}
  defp get_currency(_), do: {:error, :invalid_money}
end
```

**Domain Integration:**
```elixir
# Add to lib/moolah/finance.ex
defmodule Moolah.Finance do
  use Ash.Domain,
    otp_app: :moolah

  resources do
    resource Moolah.Finance.BudgetCategory
    resource Moolah.Finance.LifeAreaCategory
    resource Moolah.Finance.Tag
    resource Moolah.Finance.Investment
    resource Moolah.Finance.InvestmentHistory  # Add this
    resource Moolah.Finance.Operation          # Add this
    # ... other resources
  end
end
```

**Example Transaction Integration:**
```elixir
# Add to Transaction resource from Issue #5 - after_action change
change after_action(fn changeset, transaction ->
  # Only process investment transactions
  if transaction.account_id in get_investment_account_ids() or 
     transaction.target_account_id in get_investment_account_ids() do
    
    process_investment_transaction(transaction)
  end
  
  {:ok, transaction}
end)

defp process_investment_transaction(transaction) do
  case transaction.transaction_type do
    :transfer when investment_account?(transaction.target_account_id) ->
      # Money flowing INTO investment account
      create_investment_operation(%{
        type: :deposit,
        value: transaction.amount,
        investment_id: determine_target_investment(transaction),
        transaction_id: transaction.id
      })
    
    :transfer when investment_account?(transaction.account_id) ->
      # Money flowing OUT OF investment account  
      create_investment_operation(%{
        type: :withdraw,
        value: transaction.amount,
        investment_id: determine_source_investment(transaction),
        transaction_id: transaction.id
      })
    
    _ -> 
      :ok  # Non-investment transaction, no operation needed
  end
end
```

**Example Enhanced Operation Resource with Transaction Link:**
```elixir
defmodule Moolah.Finance.Operation do
  # ... existing code ...
  
  attributes do
    uuid_v7_primary_key :id
    
    attribute :type, :atom do
      constraints one_of: @types
      allow_nil? false
    end
    
    attribute :value, :money do
      allow_nil? false
      constraints storage_type: :money_with_currency
    end
    
    timestamps()
  end

  relationships do
    belongs_to :investment, Moolah.Finance.Investment do
      attribute_writable? true
      allow_nil? false
    end
    
    # Add link back to originating transaction
    belongs_to :transaction, Moolah.Finance.Transaction do
      attribute_writable? true
      allow_nil? true  # Manual operations won't have transactions
    end
  end
end
```

**Investment Selection Logic:**
When a transaction involves an investment account, the system needs to determine which specific investment receives the operation:

1. **UI Integration** (Issue #6b): User selects from dropdown of investments in that account
2. **API Integration**: Transaction payload includes `target_investment_id` field
3. **Automatic Logic**: Could default to most recent investment, or prompt user

```elixir
# Add to Transaction resource
attribute :target_investment_id, :uuid_v7 do
  # Optional - only needed for investment account transactions
end

relationships do
  # ... existing relationships ...
  
  belongs_to :target_investment, Moolah.Finance.Investment do
    attribute_writable? true
    allow_nil? true  # Only required for investment transactions
  end
end

# Validation to ensure investment belongs to target account
validate {Moolah.Finance.Validations.ValidateInvestmentAccount, []}
```

**Complete Transaction-to-Investment Flow:**
1. **User creates transaction**: "Transfer R$ 1,000 from Checking to Investment Account"
2. **System prompts**: "Which investment?" (dropdown shows investments in that account)  
3. **Transaction created**: Includes `target_investment_id`
4. **Ledger transfer created**: Money moves between accounts in double-entry system
5. **Investment operation created**: Deposit of R$ 1,000 on specific investment
6. **Investment value updated**: If user chooses to update current value
7. **History record created**: New snapshot of investment value

**Example Updated Database Schema:**
```sql
-- Add transaction reference to operations
ALTER TABLE operations ADD COLUMN transaction_id UUID REFERENCES transactions(id);

-- Add investment selection to transactions  
ALTER TABLE transactions ADD COLUMN target_investment_id UUID REFERENCES investments(id);

-- Index for transaction-operation relationships
CREATE INDEX idx_operations_transaction_id ON operations(transaction_id);
CREATE INDEX idx_transactions_target_investment_id ON transactions(target_investment_id);
```
- **deposit**: Money added to investment (increases value)
- **withdraw**: Money taken from investment (decreases value)  
- **update**: Market-driven value changes (could be positive or negative)

**History Tracking Flow:**
1. **Investment created**: History record created with initial value
2. **Investment value updated**: 
   - New history record created with new value
   - Operation record created showing the change (deposit/withdraw based on direction)
3. **Complete audit trail**: Can reconstruct investment value over time
4. **Enables accurate calculations**: True profit/loss, time-weighted returns, etc.

**Test Coverage Requirements:**
- History creation on investment creation
- History creation on investment updates  
- Operation creation on value changes
- Currency consistency validation
- No history created when value doesn't change
- Proper operation type determination (deposit vs withdraw)
- Cascade deletions work correctly
- Historical queries and aggregations

**Technical Notes:**
- Uses ash_money for proper currency handling throughout
- Automatic history creation through Ash changes
- Currency validation ensures data consistency
- Indexed for efficient time-series queries
- Foundation for future profit/loss calculations
- Maintains complete audit trail for regulatory compliance

### PR DESCRIPTION
**Pull request by JuliaMathias**
# PR: Investment Transfers, Operations, and Currency Guardrails

## Summary

This PR completes Issue 18 by wiring investment transfers and value updates into the operation/history pipeline, ensuring currency safety, and documenting/testing the flows end-to-end.

Closes #18

## What’s Included

- Added transfer-aware operation creation so investment transfers generate `:deposit`/`:withdraw` operations tied to the originating transaction.
- Introduced a `:market_update` action that records `:update` operations using the value delta, separate from cash-flow updates.
- Enforced currency consistency for investment histories and operations via dedicated validations.
- Validated `target_investment_id` usage so transfer targets are required only when an investment account is involved and must belong to the correct account.
- Linked transactions to operations (and operations to transactions) via resource changes and a new migration.
- Expanded and clarified test scenarios to describe the real-world context and expected behavior, including edge cases and error handling.
- Updated agent guidance for writing clear, meaningful test scenario comments.

## Testing

- `mix format`
- `mix precommit`

## Why This Matters

This closes the loop between cash transfers and investment tracking by ensuring operations and histories remain consistent, auditable, and tied to their source events—so future reporting and analytics stay trustworthy.

### CODE REVIEW

File: lib/moolah/finance/resources/transaction.ex [hunk 1]
```diff
@@ -47,12 +63,14 @@ defmodule Moolah.Finance.Transaction do
...
+        :target_investment_id,
         :budget_category_id,
         :life_area_category_id
       ]
 
-      change Moolah.Finance.Changes.UpdateUnderlyingTransfer
-      change Moolah.Finance.Changes.ManageTransactionTags
+      change UpdateUnderlyingTransfer
+      change ManageTransactionTags
+      change CreateInvestmentOperationFromTransaction
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/resources/transaction.ex:73]: Addressed. The change is now idempotent on update: when transfer‑relevant fields change, we delete any existing operations for the transaction and recreate a single operation with the updated values. This keeps the relationship 1:1 and prevents duplicates.

File: lib/moolah/finance/changes/create_investment_operation_from_transaction.ex [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        {:ok, record}
+      else
+        :skip -> {:ok, record}
+        {:error, error} -> {:error, error}
+      end
+    else
+      {:ok, record}
+    end
+  end
+
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/changes/create_investment_operation_from_transaction.ex]: Addressed. We now pass the changeset into the after_action, check relevant_change?/1, and reconcile operations only when transfer‑relevant fields change. This avoids extra writes on metadata‑only updates.

File: priv/repo/migrations/20260204121500_add_investment_transaction_links.exs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            column: :id,
+            name: "investment_operations_transaction_id_fkey",
+            on_delete: :nilify_all,
+            type: :uuid,
+            prefix: "public"
+          )
+    end
+
+    create index(:investment_operations, [:transaction_id])
+  end
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply priv/repo/migrations/20260204121500_add_investment_transaction_links.exs:45]: Addressed via a new migration (kept the original intact). Added a partial unique index on investment_operations.transaction_id where not null to enforce 1:1 at the DB level.

File: lib/moolah/finance/changes/create_investment_operation_from_transaction.ex [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    cond do
+      not source_is_investment and not target_is_investment ->
+        :skip
+
+      is_nil(record.target_investment_id) ->
+        {:error, "target_investment_id is required for investment transfers"}
+
+      true ->
+        {:ok, target_is_investment}
+    end
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/changes/create_investment_operation_from_transaction.ex:98]: Addressed. Missing target_investment_id now returns a structured Ash error (field: :target_investment_id) instead of a plain string.

File: lib/moolah/finance/resources/transaction.ex [hunk 5]
```diff
@@ -47,12 +63,14 @@ defmodule Moolah.Finance.Transaction do
...
       require_atomic? false
       transaction? true
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/resources/transaction.ex:75]: Addressed. Added update‑path tests that (a) reconcile the operation when amount changes and (b) avoid new operations when only description changes.

File: lib/moolah/finance/validations/validate_transaction_investment_target.ex [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  end
+
+  @spec investment_account?(Ecto.UUID.t() | nil) :: boolean()
+  defp investment_account?(nil), do: false
+
+  defp investment_account?(account_id) do
+    case Ash.get(Account, account_id) do
+      {:ok, %{account_type: :investment_account}} -> true
+      _ -> false
+    end
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/validations/validate_transaction_investment_target.ex:158]: Addressed. Account lookup failures now return explicit errors (:source_account_not_found, :target_account_not_found, :both_accounts_not_found) so missing data is surfaced clearly, and tests cover each path.

File: priv/repo/migrations/20260204121500_add_investment_transaction_links.exs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+  def down do
+    drop_if_exists index(:investment_operations, [:transaction_id])
+
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply priv/repo/migrations/20260204121500_add_investment_transaction_links.exs:49]: Addressed via a new migration that drops the redundant non‑unique index. The unique partial index remains the single index on transaction_id to avoid duplication while enforcing 1:1.
